Support (binaries) in default context in workspace

emillon opened the pull request:
This is achieved by not forcing the evaluation of the failing lazy expander if it is not required (when there is nothing to expand).

rgrinberg commented on test/blackbox-tests/test-cases/github5555.t:
```diff
@@ -0,0 +1,6 @@
+  > (env (_ (binaries x.exe)))
+  > EOF
+
+  $ dune build
```
Could you please:

* Add a description to this test:
* Split the test into a separate commit that demonstrates the failure.

hhugo commented:
It is possible to have a setup that will force this lazy. If so, I think one need to fix/hide the internal error and provide a good error message to the user.

emillon commented:
Yes, I agree. Pforms in there should still force that lazy. I'll make that a proper error message and version this behavior.

hhugo commented:
So, we don't expect `%{workspace_root}` to be working inside `dune-workspace` ?

emillon commented:
I added a description to the test. Here's the situation:

The expander that we use in the default context (`_`) does not support any pforms. And it crashes loudly when passed some. This PR addresses the second point, by displaying a (user) error message in that situation.
We can also do a bit better - for example, a string with vars that has no pforms can safely be expanded to itself - I added that as a new behavior in dune lang 3.2 for dune-workspace files.

Now there's the question of the kind of pforms we could support here. Global things like `%{workspace_root}` could work, but what about context-specific ones like `%{ocamlc}`? Should we forbid them in the default context or should they resolve to the actual context they'll be used in?
I don't think that we need to solve this right now since it seems to be enough for #5555.

rgrinberg commented on test/blackbox-tests/test-cases/github5555.t:
```diff
@@ -0,0 +1,31 @@
+  > (env (_ (binaries x%{ext_dll}.exe)))
+  > EOF
+  $ dune build
+  Error: Variables are not supported in the default context.
```
This error should really have a location. Is there a way to add it?

emillon commented:
@hhugo is the workaround in https://github.com/ocaml/dune/issues/5555#issuecomment-1103884321 (move that stanza to a the `dune` file in the same directory) good enough for you? If that's the case I think that improving the error message is good enough for now.

hhugo commented on test/blackbox-tests/test-cases/github5555.t:
```diff
@@ -0,0 +1,31 @@
+  File "dune-workspace", line 2, characters 0-26:
+  2 | (env (_ (binaries x.exe)))
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: 'env' is only available since version 3.2 of the dune language. Please
```
The error message is incorrect here. It error should mention `(env (_ (binaries ..)))` instead.

hhugo commented:
> @hhugo is the workaround in #5555 (comment) (move that stanza to a the dune file in the same directory) good enough for you? 

Not really, `dune` files scope  don't cross `dune-project` boundaries. I need to duplicate the env stanza every time I have a dune-project in my workspace.

emillon commented:
OK, I just improved the error messages. The history is a bit messy but it's fine if we squash. I can move the 3.2 creation to a separate PR since it's a bit noisy.

emillon commented:
To summarize, `(_ (binaries ...))` used to fail. Now it works provided lang is >= 3.2 AND the contents of `(binaries)` are pure text.

hhugo commented on src/dune_rules/workspace.ml:
```diff
@@ -6,10 +6,40 @@ open Dune_lang.Decoder
    simplicity *)
 let syntax = Stanza.syntax
 
+let binaries_in_any e =
```
Note that the issue is not specific to `Any`, it also triggers when a profile is specified.

``` 
(env
 (dev
   (binaries (some-path as node))))
```

rgrinberg commented:
> OK, I just improved the error messages. The history is a bit messy but it's fine if we squash. I can move the 3.2 creation to a separate PR since it's a bit noisy.

Yes, please move the version upgrade + the original test case demonstrating the bug into separate PR's.

emillon commented:
Ok, I improved the situation with better error messages and coverage for all profiles (not just the default one).
In dune < 3.2, the error only triggers when `(binaries)` is in an active config (default one or one corresponding to the active profile), so it would be a breaking change to error in that case. So I'm just emitting a warning in that case.

hhugo commented:
LGTM, thanks for the fix.

rgrinberg commented on src/dune_rules/dune_env.mli:
```diff
@@ -37,6 +37,8 @@ module Stanza : sig
     ; coq : Ordered_set_lang.Unexpanded.t
     ; format_config : Format_config.t option
+    ; error_on_use : (Loc.t * string) option
+    ; warn_on_load : (Loc.t * string) option
```
I see a couple of improvements that could be made. One is small and easy, and the other is a little more work but makes the code cleaner in the end.

1. Change string to `unit Pp.t` to reflect that this is a message
2. Change the types to `User_message.t` which is the most accurate type to use for these fields.

They are mutually exclusive, so pick one according to how much time you have.

emillon replied:
I started with `User_message.t` but this type needs an `equal` function, but `User_message` and `Pp` do not have one at the moment. I think that it's the best but it's a been more involved. Does that sound good to add `equal` functions there?

rgrinberg commented on src/dune_rules/file_binding.mli:
```diff
@@ -41,5 +41,7 @@ module Unexpanded : sig
   module L : sig
     val decode : t list Dune_lang.Decoder.t
+
+    val find_var : t list -> Loc.t option
```
Could you add a docstring to this function? The name does not make its behavior obvious.

rgrinberg commented on src/dune_rules/workspace.ml:
```diff
@@ -6,10 +6,39 @@ open Dune_lang.Decoder
 let syntax = Stanza.syntax
 
+let all_binaries e =
+  let open Dune_env.Stanza in
```
nitpick: we do not use local opens to qualify fields in the dune code base. Instead, we prefer annotating the values with the type of the record.

emillon commented:
I gave another try at writing an `equal` function for `Pp.t` but this is not great - either it operates on `Pp.t` and it needs to fail on `Format`, or on `Ast.t` but then conversion can fail.

rgrinberg commented:
>  either it operates on Pp.t and it needs to fail on Format, or on Ast.t but then conversion can fail.

Note that the limitation on Format already exists elsewhere. We already forbid its use because we require all errors to be serializable for RPC.

emillon commented:
I tried something - added a `User_message.Serializable.t` type which is just a pair (loc, string) with an equality function. What do you think?

christinerose commented on test/blackbox-tests/test-cases/github5555.t:
```diff
@@ -3,145 +3,83 @@ This test is about `binaries` in `env` stanzas in `dune-workspace` files.
+  2 | (env (_ (binaries x.exe)))
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: 'binaries' in an 'env' stanza in a dune-workspace file is only
+  available since version 3.2 of the dune language. Please update your
+  dune-project file to have (lang dune 3.2).
```
```suggestion
  available since version 3.2 of the Dune language. Please update your
  dune-project file to have (lang dune 3.2).
```

christinerose commented on test/blackbox-tests/test-cases/github5555.t:
```diff
@@ -3,145 +3,83 @@ This test is about `binaries` in `env` stanzas in `dune-workspace` files.
+  2 | (env (dev (binaries x.exe)))
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: 'binaries' in an 'env' stanza in a dune-workspace file is only
+  available since version 3.2 of the dune language. Please update your
+  dune-project file to have (lang dune 3.2).
```
```suggestion
  available since version 3.2 of the Dune language. Please update your
  dune-project file to have (lang dune 3.2).
```

christinerose commented on test/blackbox-tests/test-cases/github5555.t:
```diff
@@ -3,145 +3,83 @@ This test is about `binaries` in `env` stanzas in `dune-workspace` files.
+  2 | (env (other (binaries x.exe)))
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Warning: 'binaries' in an 'env' stanza in a dune-workspace file is only
+  available since version 3.2 of the dune language. Please update your
+  dune-project file to have (lang dune 3.2).
```
```suggestion
  available since version 3.2 of the Dune language. Please update your
  dune-project file to have (lang dune 3.2).
```

emillon replied:
This is the generic "unavailable feature" so I can't fix this one in this PR.

christinerose reviewed (changes requested):
Just a few capitalisations for Dune. The rest looks great!

emillon commented:
I rebased on top of #5634 to use the new comparison there, way better like this. Thanks!
